### DISCUSSION TITLE
Can config keys be defined multiple times?  Are previous values ignored or appended to?

### DISCUSSION THREAD

**Discussion by AGI-chandler** [fail2ban/fail2ban#3621] [Q&A] [answered]
From my cursory research, it seems python ConfigParser by default only keeps the last key/value defined, so previous values would be ignored.  Others who wanted previous values appended to appear to have written additional code surrounding ConfigParser.

I wanted to specify the `ignoreip` option in `jail.conf` multiple times, so I could include comments above each reminding why the definition exists in the first place, for example.  Is _fail2ban_ configured to allow that and append to the previous values?  If not, is there harm in adding that capability?

Thanks

**Comment by sebres** [comment, marked as answer]
> Is fail2ban configured to allow that and append to the previous values?

Only if it is specified previously in different file (e. g. in `*.conf` and need to be appended in `*.local`) or include:
```ini
ignoreip = %(known/ignoreip)s 192.0.2.1/24
```
Or in another section:
```ini
[DEFAULT]
ignoreip = ...

[jail]
ignoreip = %(default/ignoreip)s 192.0.3.1/24
```

Also one is able to specify multi-line values, e. g.:
```ini
[jail]
ignoreip = %(default/ignoreip)s 
           192.0.2.1/24
           192.0.3.1/24
           192.0.4.1/24
```
But I don't know how the interim comment line will be handled here by configreader (as part of multi-line string or as a comment), so you could check that.

> If not, is there harm in adding that capability?

The last one would be simply possible (only tokens used `splitwords`), for instance simple replacement of comments in https://github.com/fail2ban/fail2ban/blob/eed319e896f603471cc4027b353b8535faac4208/fail2ban/helpers.py#L279
as something like:
```diff
def splitwords(s):
	if not s:
		return []
+	s = re.sub(r"(?ms)^\s*[#][^\n]*\n", "", s); # remove comments
	return list(filter(bool, [v.strip() for v in re.split('[\s,]+', s)]))
```
Change of configparser module will be surely much complex.
(However I don't know whether it'd introduce some compat issues.)

Also there is `ignorecommand` which you can use to check against something else (e. g. your own script or database where you can store the reason for ignore). see https://github.com/fail2ban/fail2ban/issues/2013#issuecomment-426940189 for examples.
This way you can alter the ignore list dynamically without to reload fail2ban, do time-based whitelisting (expiration), etc.

**Reply by AGI-chandler** [reply]
Wonderful.  Thanks so much for the thorough reply!  
Can sections (and keys) be named anything?  In that case, I think I could do something like this:

```
[DEFAULT]
ignoreip = %(net1/ig)s %(net2/ig)s

# comment for net1
[net1]
ig = 1.0.0.0/8

# comment for net2
[net2]
ig = 2.0.0.0/8

...
```

I also just read in `jail.conf`: can use ';' (following a space) for inline comments.  In that case, something like this may work too:

```
ignoreip = 1.0.0.0/8     ; comment1
           2.0.0.0/8     ; comment2
           3.0.0.0/8     ; comment3
           ...
```

**Reply by sebres** [reply]
Well the first one is a bit uncomfortable as well as each new section would quasi introduce a new jail (which are disabled by default, but anyway).
As for the second, I'm unsure it would work in fail2ban - otherwise a RE with a semicolon inside would become broken (trimmed), and I never saw that before...
However related to [python documentation](https://docs.python.org/3/library/configparser.html) the variant with commented line must work in multi-line value:
```ini
[jail]
ignoreip = %(default/ignoreip)s 
           # test subnet
           192.0.2.1/24
           # still one
           192.0.3.1/24
```

**Reply by AGI-chandler** [reply]
> Well the first one is a bit uncomfortable as well as each new section would quasi introduce a new jail

Thanks I thought something like that may be the case.
<br>

> As for the second, I'm unsure it would work in fail2ban

It is [written in jail.conf](https://github.com/fail2ban/fail2ban/blob/eed319e896f603471cc4027b353b8535faac4208/config/jail.conf#L30) that way, unless i misinterpret it?
<br>

The below method is perfectly fine too, anyway...
 
> ```ini
> [jail]
> ignoreip = %(default/ignoreip)s 
>            # test subnet
>            192.0.2.1/24
>            # still one
>            192.0.3.1/24
> ```

Best,
Chandler

**Reply by sebres** [reply]
> It is [written in jail.conf](https://github.com/fail2ban/fail2ban/blob/eed319e896f603471cc4027b353b8535faac4208/config/jail.conf#L30) that way, unless i misinterpret it?

Indeed.

**Reply by eabase** [reply]
This should be more clearly added to the Wiki/FAQ pages. Very hard to find this info.
